File: packages/persistence/src/redaction/pointer.ts
/**
 * The exact RFC 6901 pointer resolution and transform order of
 * spec/redaction-semantics.md Section 3.3.1.
 *
 * The transform is a total deterministic result operation: it returns either one
 * transformed snapshot plus its canonical path list, or one structured
 * `REDACTION_POLICY_INVALID`/`REDACTION_RECEIPT_INVALID` denial. A malformed
 * pointer, missing target, normalization collision, unsupported value, or
 * internal exception never produces a partial candidate and never escapes into a
 * raw diagnostic.
 */

import type { RedactionFailureCode } from "./codes.js";
import { SECTION_11_LIMITS, type PortableLimits } from "./limits.js";
import {
  canonicalJsonString,
  compareUnicodeCodePoints,
  materializePortableJson,
  PortableJsonError,
  snapshotPortableJson,
  utf8ByteLength,
} from "./portable.js";

export const REDACTION_TOKEN = "[REDACTED]";

/**
 * Section 3.3.1 step 2: after decoding, a token equal to any of these
 * invalidates the entire transform at any depth, regardless of whether it is an
 * own JSON member.
 */
export const FORBIDDEN_POINTER_TOKENS: readonly string[] = Object.freeze([
  "__proto__",
  "prototype",
  "constructor",
]);

export type ReplacementMode = "remove" | "constant-token";

export type TransformResult =
  | {
      readonly valid: true;
      readonly output: unknown;
      readonly canonicalPaths: readonly string[];
    }
  | {
      readonly valid: false;
      readonly code: RedactionFailureCode;
      readonly reason: string;
    };

class TransformDenial extends Error {
  readonly code: RedactionFailureCode;
  readonly reason: string;

  constructor(code: RedactionFailureCode, reason: string) {
    super(`${code}: ${reason}`);
    this.name = "TransformDenial";
    this.code = code;
    this.reason = reason;
  }
}

function deny(code: RedactionFailureCode, reason: string): never {
  throw new TransformDenial(code, reason);
}

/**
 * Section 3.3.1 step 2. Split only on literal `/`; decode `~1` as `/` and `~0`
 * as `~`; reject `~` followed by anything else or by end-of-token; require the
 * decode/re-encode round trip to reproduce the input token exactly.
 */
export function decodePointer(pointer: unknown): readonly string[] {
  if (typeof pointer !== "string") deny("REDACTION_RECEIPT_INVALID", "pointer-not-a-string");
  if (pointer.length === 0) deny("REDACTION_RECEIPT_INVALID", "root-pointer-not-permitted");
  if (!pointer.startsWith("/")) deny("REDACTION_RECEIPT_INVALID", "pointer-missing-leading-slash");

  const rawTokens = pointer.slice(1).split("/");
  const tokens: string[] = [];
  for (const rawToken of rawTokens) {
    let decoded = "";
    for (let index = 0; index < rawToken.length; index += 1) {
      const character = rawToken[index] as string;
      if (character !== "~") {
        decoded += character;
        continue;
      }
      const next = rawToken[index + 1];
      if (next !== "0" && next !== "1") {
        deny("REDACTION_RECEIPT_INVALID", "invalid-escape-sequence");
      }
      decoded += next === "0" ? "~" : "/";
      index += 1;
    }
    const reEncoded = decoded.replaceAll("~", "~0").replaceAll("/", "~1");
    if (reEncoded !== rawToken) {
      deny("REDACTION_RECEIPT_INVALID", "token-is-not-canonically-escaped");
    }
    if (FORBIDDEN_POINTER_TOKENS.includes(decoded)) {
      deny("REDACTION_RECEIPT_INVALID", "forbidden-prototype-member-token");
    }
    tokens.push(decoded);
  }
  return tokens;
}

/** Encode one decoded token chain back into a canonical RFC 6901 pointer. */
export function encodePointer(tokens: readonly string[]): string {
  return tokens.map((token) => `/${token.replaceAll("~", "~0").replaceAll("/", "~1")}`).join("");
}

function enforcePointerLimits(
  pointer: string,
  tokens: readonly string[],
  limits: PortableLimits,
): void {
  if (utf8ByteLength(pointer) > limits.maxPointerUtf8Bytes) {
    deny("REDACTION_POLICY_INVALID", "pointer-exceeds-maxPointerUtf8Bytes");
  }
  if (tokens.length > limits.maxPointerTokens) {
    deny("REDACTION_RECEIPT_INVALID", "pointer-exceeds-maxPointerTokens");
  }
  for (const token of tokens) {
    if (utf8ByteLength(token) > limits.maxPointerTokenUtf8Bytes) {
      deny("REDACTION_RECEIPT_INVALID", "token-exceeds-maxPointerTokenUtf8Bytes");
    }
  }
}

function isContainerObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

const CANONICAL_ARRAY_INDEX = /^[1-9][0-9]*$/;

/**
 * Section 3.3.1 steps 3 and 4: resolve one decoded token chain against the
 * immutable snapshot. Object traversal uses only own data properties of the
 * null-prototype snapshot; a numeric-looking object key stays an object key.
 */
function resolveTokens(snapshot: unknown, tokens: readonly string[]): void {
  let current: unknown = snapshot;
  for (const token of tokens) {
    if (Array.isArray(current)) {
      if (token !== "0" && !CANONICAL_ARRAY_INDEX.test(token)) {
        deny("REDACTION_RECEIPT_INVALID", "array-token-is-not-a-canonical-index");
      }
      const parsed = Number(token);
      if (parsed >= current.length) deny("REDACTION_RECEIPT_INVALID", "array-index-out-of-range");
      current = current[parsed];
      continue;
    }
    if (isContainerObject(current)) {
      if (!Object.hasOwn(current, token)) {
        deny("REDACTION_RECEIPT_INVALID", "member-does-not-exist");
      }
      current = current[token];
      continue;
    }
    deny("REDACTION_RECEIPT_INVALID", "token-addresses-a-non-container");
  }
}

function containerAt(root: unknown, tokens: readonly string[]): unknown {
  let current: unknown = root;
  for (let index = 0; index < tokens.length - 1; index += 1) {
    const token = tokens[index] as string;
    current = Array.isArray(current)
      ? current[Number(token)]
      : (current as Record<string, unknown>)[token];
  }
  return current;
}

function isAncestor(shorter: readonly string[], longer: readonly string[]): boolean {
  if (shorter.length >= longer.length) return false;
  return shorter.every((token, index) => token === longer[index]);
}

function assignMember(target: Record<string, unknown>, key: string, value: unknown): void {
  Object.defineProperty(target, key, {
    value,
    enumerable: true,
    writable: true,
    configurable: true,
  });
}

/**
 * Apply the closed JSON Pointer transform.
 *
 * The order below is normative: Section 11 limits are evaluated before target,
 * overlap, and mutation checks, and a limit failure wins and produces no partial
 * result.
 */
export function redactionTransform(
  input: unknown,
  paths: readonly string[],
  replacementMode: string,
  limits: PortableLimits = SECTION_11_LIMITS,
): TransformResult {
  try {
    if (!Array.isArray(paths) || paths.length === 0) {
      deny("REDACTION_RECEIPT_INVALID", "at-least-one-pointer-is-required");
    }
    if (paths.length > limits.maxPointersPerRule) {
      deny("REDACTION_POLICY_INVALID", "rule-exceeds-maxPointersPerRule");
    }
    if (replacementMode !== "remove" && replacementMode !== "constant-token") {
      deny("REDACTION_RECEIPT_INVALID", "unknown-replacement-mode");
    }

    const decoded = paths.map((pointer) => {
      const tokens = decodePointer(pointer);
      enforcePointerLimits(pointer as string, tokens, limits);
      return tokens;
    });

    let snapshot: unknown;
    try {
      const captured = snapshotPortableJson(input);
      snapshot = captured.value;
      const { measurement } = captured;
      if (measurement.maxDepth > limits.maxValueDepth) {
        deny("REDACTION_RECEIPT_INVALID", "value-exceeds-maxValueDepth");
      }
      if (measurement.nodes > limits.maxValueNodes) {
        deny("REDACTION_RECEIPT_INVALID", "value-exceeds-maxValueNodes");
      }
      if (measurement.containers > limits.maxContainers) {
        deny("REDACTION_RECEIPT_INVALID", "value-exceeds-maxContainers");
      }
      if (measurement.members > limits.maxObjectMembers) {
        deny("REDACTION_RECEIPT_INVALID", "value-exceeds-maxObjectMembers");
      }
    } catch (error) {
      if (error instanceof TransformDenial) throw error;
      if (error instanceof PortableJsonError) {
        deny("REDACTION_RECEIPT_INVALID", `snapshot-${error.reason}`);
      }
      throw error;
    }

    for (let index = 1; index < paths.length; index += 1) {
      if (compareUnicodeCodePoints(paths[index - 1] as string, paths[index] as string) >= 0) {
        deny("REDACTION_RECEIPT_INVALID", "pointers-are-not-in-strict-code-point-order");
      }
    }
    for (let left = 0; left < decoded.length; left += 1) {
      for (let right = 0; right < decoded.length; right += 1) {
        if (left === right) continue;
        if (isAncestor(decoded[left] as readonly string[], decoded[right] as readonly string[])) {
          deny("REDACTION_RECEIPT_INVALID", "pointers-form-an-ancestor-descendant-pair");
        }
      }
    }

    for (const tokens of decoded) resolveTokens(snapshot, tokens);

    const locations = decoded.map((tokens) => JSON.stringify(tokens));
    if (new Set(locations).size !== locations.length) {
      deny("REDACTION_RECEIPT_INVALID", "pointers-resolve-to-duplicate-locations");
    }

    if (replacementMode === "remove") {
      for (const tokens of decoded) {
        if (Array.isArray(containerAt(snapshot, tokens))) {
          deny("REDACTION_RECEIPT_INVALID", "remove-cannot-delete-an-array-element");
        }
      }
    }

    // Deepest first; at equal depth, reverse code-point order.
    const ordered = decoded
      .map((tokens, index) => ({ tokens, pointer: paths[index] as string }))
      .sort((left, right) => {
        if (left.tokens.length !== right.tokens.length) {
          return right.tokens.length - left.tokens.length;
        }
        return compareUnicodeCodePoints(right.pointer, left.pointer);
      });

    for (const { tokens } of ordered) {
      const parent = containerAt(snapshot, tokens);
      const leaf = tokens[tokens.length - 1] as string;
      if (Array.isArray(parent)) {
        parent[Number(leaf)] = REDACTION_TOKEN;
        continue;
      }
      const record = parent as Record<string, unknown>;
      if (replacementMode === "remove") {
        delete record[leaf];
        continue;
      }
      assignMember(record, leaf, REDACTION_TOKEN);
    }

    const output = materializePortableJson(snapshot);
    const serialized = canonicalJsonString(output);
    if (utf8ByteLength(serialized) > limits.maxTransformedUtf8Bytes) {
      deny("REDACTION_RECEIPT_INVALID", "transformed-value-exceeds-maxTransformedUtf8Bytes");
    }

    return { valid: true, output, canonicalPaths: [...paths] };
  } catch (error) {
    if (error instanceof TransformDenial) {
      return { valid: false, code: error.code, reason: error.reason };
    }
    // Section 3.3.1: an internal exception never escapes into a raw diagnostic.
    return { valid: false, code: "REDACTION_RECEIPT_INVALID", reason: "internal-transform-failure" };
  }
}

File: python/src/graph_engineering/redaction/pointer.py
"""The exact RFC 6901 pointer transform frozen by Section 3.3.1.

The transform is a total deterministic result operation: it returns one
transformed snapshot plus its canonical path list, or one structured denial.  It
never returns a partly transformed object, never raises provider text, and never
mutates the caller's value.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final, Literal, TypeAlias

from ..models import JsonValue
from .errors import GuardPhase, RedactionFailure, failure
from .limits import (
    DEFAULT_LIMITS,
    PortableLimits,
    SnapshotKeyCollisionError,
    SnapshotLimitError,
    SnapshotUnsupportedValueError,
    portable_snapshot,
)

REDACTION_TOKEN: Final = "[REDACTED]"

PROTECTED_VALUE_API_VERSION: Final = (
    "graphengineering.reacher-z.github.io/protected-value/v1alpha1"
)
PROTECTED_BLOB_API_VERSION: Final = (
    "graphengineering.reacher-z.github.io/protected-blob/v1alpha1"
)

ReplacementMode: TypeAlias = Literal["remove", "constant-token"]

REPLACEMENT_MODES: tuple[ReplacementMode, ...] = ("remove", "constant-token")

# Section 3.3.1 step 2: rejected at any depth, regardless of whether the token
# names an own JSON member. RFC 6901 escapes only "~" and "/", and none of these
# names contains either character, so each has exactly one spelling.
FORBIDDEN_TOKENS: frozenset[str] = frozenset({"__proto__", "prototype", "constructor"})

_ARRAY_INDEX_ZERO: Final = "0"


class PointerSyntaxError(ValueError):
    """A pointer is not a canonical RFC 6901 pointer accepted by this contract."""


@dataclass(frozen=True, slots=True)
class PointerTransformResult:
    """A complete transform: the new snapshot plus its canonical path list."""

    output: JsonValue
    canonical_paths: tuple[str, ...]

    @property
    def count(self) -> int:
        return len(self.canonical_paths)


@dataclass(frozen=True, slots=True)
class PointerTransformDenied:
    """A total denial. No partial result exists."""

    failure: RedactionFailure


PointerOutcome: TypeAlias = PointerTransformResult | PointerTransformDenied


def decode_pointer(
    path: str,
    *,
    limits: PortableLimits = DEFAULT_LIMITS,
) -> tuple[str, ...]:
    """Decode one pointer into reference tokens, or raise :class:`PointerSyntaxError`.

    Section 11 pointer bounds are enforced here, before any target lookup.
    """

    if type(path) is not str:
        raise PointerSyntaxError("a redaction path must be a string")
    if path == "":
        # The empty pointer addresses the document root, and root replacement or
        # removal is not permitted.
        raise PointerSyntaxError("the empty pointer is not an acceptable redaction path")
    if len(path.encode("utf-8")) > limits.max_pointer_utf8_bytes:
        raise PointerSyntaxError("pointer exceeds the portable byte bound")
    if not path.startswith("/"):
        raise PointerSyntaxError("every accepted pointer starts with '/'")

    raw_tokens = path.split("/")[1:]
    if len(raw_tokens) > limits.max_pointer_tokens:
        raise PointerSyntaxError("pointer exceeds the portable reference-token bound")

    tokens: list[str] = []
    for raw in raw_tokens:
        decoded = _decode_token(raw)
        if len(decoded.encode("utf-8")) > limits.max_pointer_token_utf8_bytes:
            raise PointerSyntaxError("decoded reference token exceeds the portable byte bound")
        if _encode_token(decoded) != raw:
            raise PointerSyntaxError("reference token is not canonically escaped")
        if decoded in FORBIDDEN_TOKENS:
            raise PointerSyntaxError("pointer traverses a prototype-bearing member name")
        tokens.append(decoded)
    return tuple(tokens)


def _decode_token(raw: str) -> str:
    decoded: list[str] = []
    index = 0
    length = len(raw)
    while index < length:
        character = raw[index]
        if character != "~":
            decoded.append(character)
            index += 1
            continue
        if index + 1 >= length:
            raise PointerSyntaxError("'~' at end of a reference token is invalid")
        follower = raw[index + 1]
        if follower == "0":
            decoded.append("~")
        elif follower == "1":
            decoded.append("/")
        else:
            raise PointerSyntaxError("'~' must be followed by '0' or '1'")
        index += 2
    return "".join(decoded)


def _encode_token(decoded: str) -> str:
    # Escape "~" first, then "/", exactly as RFC 6901 requires.
    return decoded.replace("~", "~0").replace("/", "~1")


def _array_index(token: str, length: int) -> int:
    if token == _ARRAY_INDEX_ZERO:
        index = 0
    elif token and token[0] in "123456789" and token.isascii() and token.isdigit():
        index = int(token)
    else:
        raise PointerSyntaxError("array reference tokens are '0' or /[1-9][0-9]*/")
    if index >= length:
        raise PointerSyntaxError("array index is not less than the original array length")
    return index


def _resolve(snapshot: JsonValue, tokens: tuple[str, ...]) -> None:
    """Walk the snapshot iteratively; raise if any component is missing."""

    current: JsonValue = snapshot
    for token in tokens:
        kind = type(current)
        if kind is dict:
            mapping = current
            assert isinstance(mapping, dict)
            if token not in mapping:
                raise PointerSyntaxError("pointer component does not exist")
            current = mapping[token]
        elif kind is list:
            sequence = current
            assert isinstance(sequence, list)
            current = sequence[_array_index(token, len(sequence))]
        else:
            raise PointerSyntaxError("pointer traverses a non-container value")


def contains_transformed_material(snapshot: JsonValue) -> bool:
    """Detect a value that a previous attempt already transformed.

    Section 7.1 forbids re-transforming an already transformed value: hashing a
    replacement token as though it were raw yields a receipt that attests to a
    source nobody ever held, and re-protecting ciphertext produces a value whose
    plaintext is another ciphertext.
    """

    stack: list[JsonValue] = [snapshot]
    while stack:
        current = stack.pop()
        kind = type(current)
        if kind is str and current == REDACTION_TOKEN:
            return True
        if kind is list:
            assert isinstance(current, list)
            stack.extend(current)
        elif kind is dict:
            assert isinstance(current, dict)
            api_version = current.get("apiVersion")
            if api_version in (PROTECTED_VALUE_API_VERSION, PROTECTED_BLOB_API_VERSION):
                return True
            stack.extend(current.values())
    return False


def apply_pointer_transform(
    candidate: object,
    paths: object,
    replacement_mode: object,
    *,
    limits: PortableLimits = DEFAULT_LIMITS,
) -> PointerOutcome:
    """Run steps 1 through 5 of Section 3.3.1 over one candidate value."""

    denial = _denied

    if replacement_mode not in REPLACEMENT_MODES:
        return denial("replacementMode is 'remove' or 'constant-token'", phase="receipt")
    mode: ReplacementMode = replacement_mode

    # Step 1: one immutable pre-transform snapshot, portable-string normalized,
    # with normalized-key collisions rejected before any path is consulted.
    try:
        snapshot, _ = portable_snapshot(candidate, limits=limits)
    except SnapshotKeyCollisionError:
        return denial("object keys collide after portable normalization", phase="snapshot")
    except SnapshotLimitError:
        return denial("candidate exceeds a portable resource bound", phase="snapshot")
    except SnapshotUnsupportedValueError:
        return denial("candidate is not portable JSON", phase="snapshot")

    if type(paths) is not tuple and type(paths) is not list:
        return denial("paths must be an ordered pointer list", phase="receipt")
    path_list = list(paths)
    if not path_list:
        return denial("a transform requires at least one path", phase="receipt")

    # Section 3.3.1: Section 11 limits are evaluated before target existence and
    # overlap checks, and produce no partial result.
    if len(path_list) > limits.max_pointers_per_rule:
        return denial("path count exceeds the portable bound", phase="receipt")

    # Step 2: syntax, escaping round-trip, and prototype rejection.
    decoded: list[tuple[str, ...]] = []
    for path in path_list:
        try:
            decoded.append(decode_pointer(path, limits=limits))
        except PointerSyntaxError:
            return denial("pointer is not a canonical safe RFC 6901 pointer", phase="receipt")

    # Step 4: strict canonical order, duplicate source strings, duplicate
    # resolved locations, and ancestor/descendant overlap.
    for index in range(1, len(path_list)):
        if not path_list[index - 1] < path_list[index]:
            return denial("paths must be in strictly increasing code-point order", phase="receipt")

    resolved: set[tuple[str, ...]] = set()
    for tokens in decoded:
        if tokens in resolved:
            return denial("duplicate resolved pointer location", phase="receipt")
        resolved.add(tokens)
    for tokens in decoded:
        for other in decoded:
            if tokens is other:
                continue
            if len(tokens) < len(other) and other[: len(tokens)] == tokens:
                return denial("overlapping ancestor/descendant pointers", phase="receipt")

    # Step 3 and step 4 target existence, against the one immutable snapshot.
    for tokens in decoded:
        try:
            _resolve(snapshot, tokens)
        except PointerSyntaxError:
            return denial("pointer target does not exist", phase="receipt")

    if mode == "remove":
        for tokens in decoded:
            if not _removes_object_member(snapshot, tokens):
                return denial("remove addresses the root or an array element", phase="receipt")

    # Step 5: apply from greatest depth to least, reverse code-point order for
    # equal depth. A fresh copy keeps the pre-transform snapshot immutable.
    # ``path_list`` is already strictly increasing, so a greater position is a
    # greater pointer and ``-position`` is exactly reverse code-point order.
    output, _ = portable_snapshot(snapshot, limits=limits)
    order = sorted(
        range(len(decoded)),
        key=lambda position: (-len(decoded[position]), -position),
    )
    for position in order:
        _apply_one(output, decoded[position], mode)

    # Step 6 (first half): the result must still be portable JSON within bounds.
    try:
        result, _ = portable_snapshot(output, limits=limits)
    except (SnapshotKeyCollisionError, SnapshotLimitError, SnapshotUnsupportedValueError):
        return denial("transformed value is not portable JSON", phase="receipt")

    return PointerTransformResult(output=result, canonical_paths=tuple(path_list))


def _removes_object_member(snapshot: JsonValue, tokens: tuple[str, ...]) -> bool:
    if not tokens:
        return False
    parent: JsonValue = snapshot
    for token in tokens[:-1]:
        if type(parent) is dict:
            assert isinstance(parent, dict)
            parent = parent[token]
        else:
            assert isinstance(parent, list)
            parent = parent[_array_index(token, len(parent))]
    return type(parent) is dict


def _apply_one(root: JsonValue, tokens: tuple[str, ...], mode: ReplacementMode) -> None:
    parent: JsonValue = root
    for token in tokens[:-1]:
        if type(parent) is dict:
            assert isinstance(parent, dict)
            parent = parent[token]
        else:
            assert isinstance(parent, list)
            parent = parent[_array_index(token, len(parent))]
    last = tokens[-1]
    if type(parent) is dict:
        assert isinstance(parent, dict)
        if mode == "remove":
            del parent[last]
        else:
            parent[last] = REDACTION_TOKEN
        return
    assert isinstance(parent, list)
    parent[_array_index(last, len(parent))] = REDACTION_TOKEN


def _denied(_reason: str, *, phase: GuardPhase = "receipt") -> PointerTransformDenied:
    # The reason string is developer documentation at the call site and is never
    # persisted: Section 10 forbids echoing the offending value, and a pointer or
    # key name is caller-derived material.
    return PointerTransformDenied(failure("REDACTION_RECEIPT_INVALID", phase))
